lib/utils/tls.go

```diff
@@ -54,6 +67,71 @@ type CertificateName struct {
 	OrganizationalUnit []string `json:"org_unit"`
 }
 
+// GenerateSelfSignedCert generates a self signed certificate that
+// is valid for given domain names and ips, returns PEM-encoded bytes with key and cert
+// Generates a certificate that is compatible with the MacOS requirements described at:
+// https://support.apple.com/en-us/HT210176
+func GenerateSelfSignedCert(hostNames []string) (*teleutils.TLSCredentials, error) {
+	priv, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	notBefore := time.Now()
+	notAfter := notBefore.Add(time.Hour * 24 * 825) // 825 days or fewer is the required validity period for MacOS
+
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	entity := pkix.Name{
+		CommonName:   "localhost",
+		Country:      []string{"US"},
+		Organization: []string{"localhost"},
+		// OrganizationalUnit is needed in order to be able to identify the cert when doing
+		// automated cert rotation. If a user decides to use their own cert
+		// we should not rotate.
+		OrganizationalUnit: []string{SelfSignedCertOrg},
+	}
+
+	template := x509.Certificate{
+		SerialNumber:          serialNumber,
+		Issuer:                entity,
+		Subject:               entity,
+		NotBefore:             notBefore,
+		NotAfter:              notAfter,
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth}, // MacOS specific requirement
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+	}
+
+	// collect IP addresses localhost resolves to and add them to the cert. template:
+	template.DNSNames = append(hostNames, "localhost.local")
+	ips, _ := net.LookupIP("localhost")
```

Comment:
Need to be careful of a very particular oddity of golang and some base images missing /etc/nsswitch.conf. I checked and at least on my test cluster we're missing the nsswitch configuration.

Several versions of golang when missing nsswitch configuration, will use DNS without local host resolution. This means this lookup could query external DNS servers and be subject to a number of unexpected results.

Just one of those very peculiar quirks of go. I would consider just hard coding the response to 127.0.0.1 instead of relying on name resolution of the localhost name.